src/TurnByTurn/RoutingSample.Desktop/MainWindow.xaml.cs

```diff
@@ -17,14 +18,26 @@ public partial class MainWindow : Window
 
         public MainWindow()
         {
+            // Create an API key at https://developers.arcgis.com/api-keys/, configure it with routing and geocoding services, then paste below.
+            Esri.ArcGISRuntime.ArcGISRuntimeEnvironment.ApiKey = "";
             InitializeComponent();
 
+            CheckKey();
+
             _mainViewModel = new MainViewModel();
             _mainViewModel.LocationDisplay = MapView.LocationDisplay;
             _mainViewModel.LocationDisplay.NavigationPointHeightFactor = 0.5;
 
             DataContext = _mainViewModel;
         }
+        private void CheckKey()
+        {
+            if (string.IsNullOrEmpty(Esri.ArcGISRuntime.ArcGISRuntimeEnvironment.ApiKey))
+            {
+                MessageBox.Show("See MainWindow.xaml.cs for info about setting API key", "Error - No API Key provided", MessageBoxButton.OK, MessageBoxImage.Warning);
+                System.Environment.Exit(0);
+            }
+        }
```

Comment:
This wouldn't be necessary if you hadn't removed the login window. Signing into the portal gives you the API key needed.

Reply:
Using the API key seemed less annoying than requiring login, especially since the existing GenerateTokenCredential implementation is case-sensitive for usernames, whereas usernames are not case sensitive on the rest of the platform. I spent quite a lot of time trying to figure out why it was broken before remembering this quirk.

The alternative is using OAuth, but that requires even more configuration than setting an API key.